### PR TITLE
SPDX 2.2 Files Validator

### PR DESCRIPTION
**Pull request by aasim**
Validates a SPDX 2.2 file against a drop

### CODE REVIEW

File: src/Microsoft.Sbom.Api/Converters/IManifestPathConverter.cs [hunk 1]
```diff
@@ -11,6 +11,6 @@ public interface IManifestPathConverter
         /// </summary>
         /// <param name="path">The relative path of the file.</param>
         /// <returns>The file path in the manifest format and boolean for if the path is outside the BuildDropPath.</returns>
-        (string, bool) Convert(string path);
+        (string, bool) Convert(string path, bool prependDotToPath = false);
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Api/Converters/IManifestPathConverter.cs:14]: Add documentation of the parameter

File: src/Microsoft.Sbom.Api/Entities/ErrorType.cs [hunk 2]
```diff
@@ -38,6 +38,9 @@ public enum ErrorType
         JsonSerializationError = 8,
 
         [EnumMember(Value = "Unsupported hash algorithm")]
-        UnsupportedHashAlgorithm = 9
+        UnsupportedHashAlgorithm = 9,
+
+        [EnumMember(Value = "Referenced SBOM file")]
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Api/Entities/ErrorType.cs:43]: Why is this an error?

Comment (aasim) [reply src/Microsoft.Sbom.Api/Entities/ErrorType.cs:43]: This is a referenced SBOM, which can sometimes be outside of the drop and not present in the drop. So, we skip referenced SBOMs from validation.

File: src/Microsoft.Sbom.Api/Converters/IManifestPathConverter.cs (hunk 1 above)
Comment (aasim) [reply src/Microsoft.Sbom.Api/Converters/IManifestPathConverter.cs:14]: Fixed

File: src/Microsoft.Sbom.Contracts/ISBOMValidator.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        /// <param name="algorithmName">The algorithm to be used to validate checksums. Default value is SHA256.</param>
+        /// <param name="specifications">A list of SBOM formats to validate.</param>
+        /// <param name="manifestDirPath">If provided will use this value for the SBOM. Default will search for the _manifest directory in the build drop path.</param>
+        /// <param name="catalogFilePath">Optional path of the catalog file used to verify the SBOM.</param>
+        /// <param name="validateSignature">If true, we will try to validate the signature of the SBOM using the catalog file.</param>
+        /// <param name="ignoreMissing">If true, we will ignore reporting files missing from the disk as errors.</param>
+        /// <param name="rootPathFilter">If you're downloading only a part of the drop using the '-r' or 'root' parameter in the
+        /// drop client, specify the same string value here in order to skip validating paths that are not downloaded.</param>
+        /// <param name="runtimeConfiguration">Additional parameters to configure the SBOM .</param>
```
Comment (abdulkkMS) [inline src/Microsoft.Sbom.Contracts/ISBOMValidator.cs:30]: Will this in future return a ValidationResult Object?

File: src/Microsoft.Sbom.Api/Converters/IManifestPathConverter.cs [hunk 4]
```diff
@@ -10,7 +10,8 @@ public interface IManifestPathConverter
         /// that the manifest implements.
         /// </summary>
         /// <param name="path">The relative path of the file.</param>
+        /// <param name="prependDotToPath">If true we will prepend a . before the path.</param>
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Api/Converters/IManifestPathConverter.cs:13]: >prependDotToPath [](http://example.com/codeflow?start=25&length=16)

Why we need to pre-append a dot?

File: src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    /// Validates hashes from multiple file locations (on disk and inside SBOM) simulatenously using
+    /// a conncurrent dictionary.
+    /// </summary>
+    public class ConcurrentHashValidator
+    {
+        private readonly FileHashesDictionary fileHashesDictionary;
+
+        public ConcurrentHashValidator(FileHashesDictionary fileHashesDictionary)        
+        {
+            this.fileHashesDictionary = fileHashesDictionary;
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs:25]: >fileHashesDictionary [](http://example.com/codeflow?start=40&length=20)

Check for null

File: src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            var sha256Checksum = internalFileInfo.Checksum.Where(c => c.Algorithm == AlgorithmName.SHA256).FirstOrDefault();
+            var fileHashes = new FileHashes();
+            fileHashes.SetHash(internalFileInfo.FileLocation, sha256Checksum);
+            FileValidationResult failureResult = null;
+
+            var newValue = fileHashesDictionary.FileHashes.AddOrUpdate(internalFileInfo.Path, fileHashes, (key, oldValue) =>
+            {
+                // This means a file with the same location was already added to the dictionary.
+                if (oldValue.GetHash(internalFileInfo.FileLocation) != null)
+                {
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs:59]: When can we find this case?

File: src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+using System.Threading.Channels;
+using System.Threading.Tasks;
+
+namespace Microsoft.Sbom.Api.Executors
+{
+    /// <summary>
+    /// Validates hashes from multiple file locations (on disk and inside SBOM) simulatenously using
+    /// a conncurrent dictionary.
+    /// </summary>
+    public class ConcurrentHashValidator
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs:19]: >ConcurrentHashValidator [](http://example.com/codeflow?start=17&length=23)

Should this be ConcurrentHash256Validator?

File: src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            if (failureResult != null)
+            {
+                await errors.Writer.WriteAsync(failureResult);
+                return;
+            }
+
+            // If we have the files from both locations present in the hash, validate if the hashes match.
+            if (newValue.FileLocation == Sbom.Entities.FileLocation.All)
+            {
+                if (string.Equals(newValue.OnDiskHash.ChecksumValue, newValue.SBOMFileHash.ChecksumValue, StringComparison.InvariantCultureIgnoreCase))
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs:82]: >InvariantCultureIgnoreCase [](http://example.com/codeflow?start=123&length=26)

Should this be OrdinalComparison?

File: src/Microsoft.Sbom.Api/Executors/EnumeratorChannel.cs [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    /// <summary>
+    /// A executor that enumerates over objects in an enumerator.
+    /// </summary>
+    public class EnumeratorChannel
+    {
+        private readonly ILogger log;
+
+        public EnumeratorChannel(ILogger log)
+        {
+            this.log = log;
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Api/Executors/EnumeratorChannel.cs:22]: >log [](http://example.com/codeflow?start=23&length=3)

Check for null

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Validator.cs [hunk 10]
```diff
@@ -14,17 +15,17 @@ public class Validator : IManifestInterface
...
             Name = Constants.SPDXName,
             Version = Constants.SPDXVersion
         };
 
-        public ManifestData ParseManifest(string manifest)
-        {
-            throw new NotImplementedException($"Currently we don't support validating SPDX 2.2 SBOMs");
-        }
+        public ManifestData ParseManifest(string manifest) 
+            => throw new NotImplementedException($"Currently we don't support parsing complete SPDX 2.2 SBOMs");
```
Comment (edgarrs) [inline src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Validator.cs:25]: >Currently we don't support parsing complete SPDX 2.2 SBOMs" [](http://example.com/codeflow?start=51&length=59)

What does this mean?

**Review by edgarrs — APPROVED** [review]
LGTM but take a look at my comments

File: src/Microsoft.Sbom.Contracts/ISBOMValidator.cs (hunk 3 above)
Comment (aasim) [reply src/Microsoft.Sbom.Contracts/ISBOMValidator.cs:30]: Yeah, this is the first phase, we have some additional validations we will add and change this behavior. 

Currently, you can read the output as its written to a file to check the results.

File: src/Microsoft.Sbom.Api/Converters/IManifestPathConverter.cs (hunk 4 above)
Comment (aasim) [reply src/Microsoft.Sbom.Api/Converters/IManifestPathConverter.cs:13]: This is how SPDX file paths are stored by default, we add a . to the relative path, so we have to do the same internally for it to match.

File: src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs (hunk 6 above)
Comment (aasim) [reply src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs:59]: If I have a SBOM that has two entries for the same file path, we will hit this error.

File: src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Validator.cs (hunk 10 above)
Comment (aasim) [reply src/Microsoft.Sbom.Parsers.Spdx22SbomParser/Validator.cs:25]: It means that we don't support parsing the whole SBOM file at once... I will update the string to be a little better

File: src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs (hunk 7 above)
Comment (aasim) [reply src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs:19]: fixed

File: src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs (hunk 5 above)
Comment (aasim) [reply src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs:25]: fixed

File: src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs (hunk 8 above)
Comment (aasim) [reply src/Microsoft.Sbom.Api/Executors/ConcurrentHashValidator.cs:82]: This is a utf 8 string, we convert it from binary to a string representation to serialize to file.

File: src/Microsoft.Sbom.Api/Executors/EnumeratorChannel.cs (hunk 9 above)
Comment (aasim) [reply src/Microsoft.Sbom.Api/Executors/EnumeratorChannel.cs:22]: fixed

**Pull request merged**